### PR TITLE
[backend/client] Make x_mitre_id case-insensitive for standard_id generation (#16132)

### LINKED ISSUE #16132: [Backend] Attack Pattern / Course of Action standard_id generation is case-sensitive on x_mitre_id
﻿## Description

When generating the `standard_id` of an Attack Pattern (or Course of Action) from its `x_mitre_id` (also accepted as `x_opencti_external_id` in STIX bundles), the value is currently treated case-sensitively. As a consequence, two ingestions of the same MITRE technique with different casing (e.g. `T1059` and `t1059`) end up creating two distinct entities instead of being deduplicated.

This is inconsistent with other entities such as Vulnerability (whose `name` is normalized via `normalizeName`) or Software (made case-insensitive for `name` in #15633).

## Environment

1. OS (where OpenCTI server runs): N/A
2. OpenCTI version: master / 7.260521.0
3. OpenCTI client: backend (graphql) + python client
4. Other environment details: N/A

## Reproducible Steps

1. Create an Attack Pattern with `x_mitre_id: 'T1059'`.
2. Re-import the same Attack Pattern with `x_mitre_id: 't1059'` (different casing).
3. A second Attack Pattern is created instead of being merged with the existing one.

## Expected Output

The two attack patterns should resolve to the same `standard_id` and be deduplicated/merged.

## Actual Output

Two distinct entities are created, each with its own `standard_id`.

## Additional information

- Reference code: `opencti-platform/opencti-graphql/src/schema/identifier.js` — the `x_mitre_id` field has no resolver, so the default `.trim()` branch is used and case is preserved.
- Same problem applies to Course of Action which shares the same id rule `[[{ src: X_MITRE_ID_FIELD }], [{ src: NAME_FIELD }]]`.
- Fix: add an `x_mitre_id` resolver that calls `normalizeName` (lowercase + trim), mirror in `client-python` (`opencti_attack_pattern.py` and `opencti_course_of_action.py`), and add a migration that merges existing duplicates and rewrites the `standard_id` of remaining entities — mirroring the approach used in migration `1776859399668-software_case_insensitive_standard_id.js`.

### PR DESCRIPTION
**Pull request by SamuelHassine**
### Proposed changes

* Make `x_mitre_id` case-insensitive (lowercase + trim) when computing the `standard_id` of **Attack Pattern** and **Course of Action**, by adding an `x_mitre_id` resolver in `stixBaseEntityContribution.resolvers` (`opencti-platform/opencti-graphql/src/schema/identifier.js`). Re-importing the same MITRE technique (e.g. `T1059` and then `t1059`, or with extra whitespace) now deduplicates the entity instead of creating a second one, consistently with how `name` is normalized for vulnerabilities and how `name` was made case-insensitive for Software in #15633.
* Mirror the same normalization in the Python client (`AttackPattern.generate_id` / `CourseOfAction.generate_id`) so STIX bundles produced by pycti directly compute the same canonical `standard_id` as the backend. When the normalized `x_mitre_id` is empty (whitespace-only input), the python client now also falls back to `name` to mirror the backend's `filteredIdContributions` empty-field filtering, so client and server stay aligned even on the edge case.
* Add migration `1779435787999-attack_pattern_case_insensitive_standard_id.js`, modeled on `1776859399668-software_case_insensitive_standard_id.js`: it recomputes the new `standard_id` for every Attack Pattern and Course of Action, merges colliding duplicates into a single canonical target via `mergeEntities` (the merge target is picked by relation count DESC — i.e. the most-connected entity — to minimize relation rewriting, ties broken by oldest `created_at` then lowest `internal_id`), and bulk-rewrites the remaining singletons' `standard_id` while archiving the previous one in `x_opencti_stix_ids` so older references keep resolving. Progress logs go through `logMigration.info` so the operator can follow the merge and bulk-rewrite loops from the migration log channel.
* Update the existing unit tests (`identifier-test.js` and `client-python` `test_bundle_ids_rewrite.py`) for the new canonical UUIDs of `MITREID` / `T1368` inputs, and add new assertions that explicitly cover case-insensitivity, surrounding-whitespace trimming, and the whitespace-only fallback to `name` for both Attack Pattern and Course of Action — locking client/server id parity in both test suites.

### Related issues

* Closes #16132

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case (coverage and e2e)
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

The migration intentionally has no `down` step: previous `standard_id`s are preserved inside `x_opencti_stix_ids` so no data is lost, but recomputing the pre-merge case-sensitive ids would require the original pre-merge state which is no longer available after merges. This follows the exact same convention as the Software migration (#15633).

### CODE REVIEW

File: opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      logApp.error(`${message} > failed to merge group for ${newId}`, { cause: err, targetId: target.internal_id, sourceIds: sources.map((s) => s.internal_id) });
+    }
+  }
+
+  // Bulk-rewrite the standard_id of all remaining (non-colliding) entities whose id actually
+  // changed. The previous standard_id is archived in x_opencti_stix_ids so older STIX IDs
+  // keep resolving to the same entity.
+  const bulkOperations = [];
+  const singletonGroups = groups.filter((g) => g.length === 1);
+  for (let index = 0; index < singletonGroups.length; index += 1) {
```
Comment (aHenryJard) [inline opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js:102]: Could you please add a log to have the progress on loops (like processing element N/total)?

File: opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    );
+  }
+
+  if (bulkOperations.length > 0) {
+    let currentProcessing = 0;
+    const groupsOfOperations = R.splitEvery(MAX_BULK_OPERATIONS, bulkOperations);
+    const concurrentUpdate = async (bulk) => {
+      await elBulk(context, { refresh: true, timeout: BULK_TIMEOUT, body: bulk });
+      currentProcessing += bulk.length;
+      logApp.info(`${message} > bulk rewrote ${entityType} standard ids: ${currentProcessing} / ${bulkOperations.length}`);
```
Comment (aHenryJard) [inline opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js]: you need to use logMigration.info
```suggestion
      logMigration.info(`${message} > bulk rewrote ${entityType} standard ids: ${currentProcessing} / ${bulkOperations.length}`);
```

File: opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js (hunk 1 above)
Comment (aHenryJard) [reply opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js:102]: ok I think progress is log later on the code, but please make sure that it's possible to follow progress to estimate remaining time.

Comment (SamuelHassine) [reply opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js:102]: Good catch. The merge loop and the bulk loop both already log `(N/total)`, but they were going through `logApp` so they were not visible in the migration log channel. I've routed both progress lines through `logMigration.info` in 2b87c68 so the operator can follow progress and estimate remaining time from the migration logs themselves, while the failure path keeps using `logApp.error` for app-level alerting.

File: opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js (hunk 2 above)
Comment (SamuelHassine) [reply opencti-platform/opencti-graphql/src/migrations/1779435787999-attack_pattern_case_insensitive_standard_id.js]: Done in 2b87c68 — switched both the bulk-rewrite progress line and the merge-loop progress line to `logMigration.info`. Errors stay on `logApp.error`.

**Comment by SamuelHassine** [comment]
## Review and fix pass — summary

I ran a full end-to-end pass on this PR: independent senior-reviewer code read of all touched files (not just the diff), addressing the existing review comments, and CI verification.

### Changes pushed (`2b87c68`)

- Routed both the merge-loop progress line and the bulk-rewrite progress line in `1779435787999-attack_pattern_case_insensitive_standard_id.js` through `logMigration.info` instead of `logApp.info`. This is what @aHenryJard asked for in the two review threads: progress is now visible in the migration log channel so the operator can follow it and estimate remaining time. Failure paths keep `logApp.error` for app-level alerting.

### Review threads

Both review threads from @aHenryJard have been replied to and resolved (`PRRT_kwDOCar6qs6EHoOP`, `PRRT_kwDOCar6qs6EHqro`).

### Independent review findings

I also did a fresh file-by-file read of `migrations/1779435787999-...`, `schema/identifier.js`, both pycti `generate_id` callsites, and all the test data updates. Everything else looked correct:

- `x_mitre_id` resolver applies `normalizeName` consistently with `name`, `identity_class`, `value`; defensive `(name || '').toLowerCase().trim()` handles undefined/null.
- Migration matches the conventions established by the Software case-insensitive migration (`1776859399668-...`), with one small improvement: the merge target is picked by relation count DESC (richest, most-connected) instead of just by `created_at`, so we minimize relation rewriting and keep the most-connected node canonical. Ties broken deterministically.
- `elBatchIdsWithRelCount` is the right primitive for that and is only called for colliding entities (singletons skip the rel-count load).
- `patchAttribute` is correctly called before `mergeEntities` so the previous `standard_id`s are archived into `x_opencti_stix_ids` by the middleware.
- Singletons whose id changed are bulk-rewritten with their old `standard_id` archived in `x_opencti_stix_ids`, so older STIX IDs keep resolving.
- Filename timestamp `1779435787999` is chronologically after the most recent existing migration (`1779117914543-fix-playbook-config-bundle-scope.js`).
- Python `lower().strip()` matches the JS `(name || '').toLowerCase().trim()` semantics, so pycti-generated bundles compute the same canonical `standard_id` as the backend.
- Test updates are internally consistent (every cascading UUID change is mirrored across `identifier-test.js`, `filterGroup-test.js`, `middleware-test.js`, `stix-test.js`, `attackPattern-test.js`, `container-test.ts`, `courseOfAction-test.js`, `stream-event-stix2-reports.json`, and `test_bundle_ids_rewrite.py`).

### CI

All 50+ required checks are green on `2b87c68`: GitHub Actions, CircleCI, codecov, CodeQL, signed commits, PR title convention, linked issue, License check, frontend unit + e2e, backend unit + integration + rules-and-taxii, pycti 3.10 / 3.11 / 3.12. CLA shows "not required (organization member)".

### Remaining blocker

`mergeStateStatus: BLOCKED` because `reviewDecision: REVIEW_REQUIRED`. I did not run `gh pr review --approve` because I'm the PR author and GitHub disallows self-approval — this PR still needs a maintainer approval to merge.

File: client-python/pycti/entities/opencti_attack_pattern.py [hunk 3]
```diff
@@ -282,7 +282,7 @@ def generate_id(name, x_mitre_id=None):
         :rtype: str
         """
         if x_mitre_id is not None:
-            data = {"x_mitre_id": x_mitre_id.strip()}
+            data = {"x_mitre_id": x_mitre_id.lower().strip()}
         else:
             data = {"name": name.lower().strip()}
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply client-python/pycti/entities/opencti_attack_pattern.py]: Good catch, fixed in d26e5a9. `AttackPattern.generate_id` now normalizes `x_mitre_id` first and treats it as unset when the normalized value is empty, falling back to `name` exactly like the backend does in `filteredIdContributions` (where empty contributing fields are filtered out and the next way — `name` — is tried). Added an explicit assertion in `test_bundle_ids_rewrite.py` (`x_mitre_id: "   "`, `name: "attack"` → same id as `name: "attack"` alone) plus a mirror assertion in `identifier-test.js`, so client/server parity for whitespace-only `x_mitre_id` is now locked down in both test suites.

File: client-python/pycti/entities/opencti_course_of_action.py [hunk 4]
```diff
@@ -256,7 +256,7 @@ def generate_id(name, x_mitre_id=None):
         :rtype: str
         """
         if x_mitre_id is not None:
-            data = {"x_mitre_id": x_mitre_id.strip()}
+            data = {"x_mitre_id": x_mitre_id.lower().strip()}
         else:
             data = {"name": name.lower().strip()}
         data = canonicalize(data, utf8=False)
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply client-python/pycti/entities/opencti_course_of_action.py]: Same fix applied here in d26e5a9. `CourseOfAction.generate_id` now mirrors the backend: it normalizes `x_mitre_id` and falls back to `name` when the normalized value is empty, so a whitespace-only `x_mitre_id` produces the same canonical id on the client as the backend would compute (the backend filters the empty `x_mitre_id` out in `filteredIdContributions` and resolves via the `name` way). Verified by a new assertion in `test_bundle_ids_rewrite.py` and a parity assertion in `identifier-test.js`.

File: opencti-platform/opencti-graphql/tests/01-unit/domain/identifier-test.js [hunk 5]
```diff
@@ -178,6 +178,25 @@ describe('identifier', () => {
...
+    const apPadded = generateStandardId(ENTITY_TYPE_ATTACK_PATTERN, { x_mitre_id: '  T1059  ' });
+    expect(apLower).toEqual(apUpper);
+    expect(apPadded).toEqual(apUpper);
+    // The name fallback must remain unaffected when x_mitre_id is set
+    const apWithName = generateStandardId(ENTITY_TYPE_ATTACK_PATTERN, { x_mitre_id: 'T1059', name: 'Command and Scripting Interpreter' });
+    expect(apWithName).toEqual(apUpper);
+
+    const coaUpper = generateStandardId(ENTITY_TYPE_COURSE_OF_ACTION, { x_mitre_id: 'M1056' });
+    const coaLower = generateStandardId(ENTITY_TYPE_COURSE_OF_ACTION, { x_mitre_id: 'm1056' });
+    expect(coaLower).toEqual(coaUpper);
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply opencti-platform/opencti-graphql/tests/01-unit/domain/identifier-test.js:198]: Good catch — the Course of Action padded-whitespace assertion was indeed missing. Added in d26e5a9: `coaPadded = generateStandardId(ENTITY_TYPE_COURSE_OF_ACTION, { x_mitre_id: '  M1056  ' })` is now asserted equal to `coaUpper`, so the trimming behaviour is explicitly locked for both Attack Pattern and Course of Action. Also added a separate pair of assertions in the same test verifying that a whitespace-only `x_mitre_id` falls back to `name` for both types (matching the python client behaviour and the `filteredIdContributions` empty-field filtering documented inline).

File: client-python/tests/01-unit/stix/test_bundle_ids_rewrite.py [hunk 6]
```diff
@@ -40,8 +42,10 @@ def test_ids_generation():
     # report
     assert gen_id({"type": "report", "name": "Report", "published": "2022-11-25T19:00:05.000Z"}) == "report--761c6602-975f-5e5e-b220-7a2d41f33ce4"
     # course-of-action
-    assert gen_id({"type": "course-of-action", "x_mitre_id": "MITREID"}) == "course-of-action--b74cfee2-7b14-585e-862f-fea45e802da9"
-    assert gen_id({"type": "course-of-action", "x_mitre_id": "MITREID", "name": "Name"}) == "course-of-action--b74cfee2-7b14-585e-862f-fea45e802da9"
+    assert gen_id({"type": "course-of-action", "x_mitre_id": "MITREID"}) == "course-of-action--66488b4f-c9d1-5e36-99cc-372b34ebeeee"
+    assert gen_id({"type": "course-of-action", "x_mitre_id": "MITREID", "name": "Name"}) == "course-of-action--66488b4f-c9d1-5e36-99cc-372b34ebeeee"
+    # x_mitre_id is case-insensitive: 'MITREID' and 'mitreid' must resolve to the same standard id
+    assert gen_id({"type": "course-of-action", "x_mitre_id": "mitreid"}) == "course-of-action--66488b4f-c9d1-5e36-99cc-372b34ebeeee"
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply client-python/tests/01-unit/stix/test_bundle_ids_rewrite.py:53]: Added in d26e5a9 — the test file now explicitly asserts both the trimming behaviour and the empty-fallback behaviour for both entity types: `gen_id({"type": "attack-pattern", "x_mitre_id": "  MITREID  "})` and `gen_id({"type": "course-of-action", "x_mitre_id": "  MITREID  "})` both equal the `MITREID` id, and `gen_id({"type": "attack-pattern", "x_mitre_id": "   ", "name": "attack"})` / `gen_id({"type": "course-of-action", "x_mitre_id": "   ", "name": "Name"})` both fall back to the `name`-only id. Combined with the matching JS assertions in `identifier-test.js`, client and backend now share the same canonical id for surrounding whitespace and for the whitespace-only edge case.

**Comment by SamuelHassine** [comment]
## Review and fix pass — summary (second iteration)

Full end-to-end pass on the four new Copilot review comments + a fresh senior-reviewer read of all touched files.

### Changes pushed (`d26e5a9`)

- `client-python/pycti/entities/opencti_attack_pattern.py` and `opencti_course_of_action.py`: `generate_id` now normalizes `x_mitre_id` first and treats it as unset when the normalized value is empty (whitespace-only input), falling back to `name`. This mirrors the backend behaviour in `filteredIdContributions` (which filters out empty contributing fields and tries the next way, i.e. `name`), so pycti-generated bundles and the backend now compute the same canonical `standard_id` even on the whitespace-only edge case.
- `opencti-platform/opencti-graphql/tests/01-unit/domain/identifier-test.js`: added the missing Course of Action padded-whitespace assertion (`'  M1056  '` → same id as `'M1056'`) plus a pair of parity assertions showing that a whitespace-only `x_mitre_id` falls back to `name` for both entity types.
- `client-python/tests/01-unit/stix/test_bundle_ids_rewrite.py`: added padded-value assertions for both `attack-pattern` and `course-of-action` (`'  MITREID  '` → same id as `'MITREID'`) and explicit empty-fallback assertions, so client/server parity for surrounding whitespace and for blank `x_mitre_id` is locked down in both test suites.

### Review threads

All 6 review threads on this PR are now resolved (0 unresolved): the two earlier `@aHenryJard` migration-log threads from the previous pass, plus the four new `@Copilot` threads addressed by this commit.

### Independent senior-reviewer findings

A fresh file-by-file re-read confirmed the previous pass's conclusions, with one new finding fixed:

- New must-fix → applied: empty/whitespace `x_mitre_id` was not handled symmetrically between pycti and the backend. Fixed in `d26e5a9`, locked down in both test suites.
- Identifier resolver, migration target-picking, `patchAttribute` → `mergeEntities` ordering, `x_opencti_stix_ids` archival, migration timestamp, log channel routing → all still correct as reviewed in the previous pass.
- Lint / format: `black --check`, `isort --check-only`, `flake8 --ignore=E,W` pass locally on all modified python files. JS unit suite couldn't be executed locally (Windows-only `nodecallspython.node` build issue), but the new assertions only exercise pre-existing production code paths (`normalizeName` + `filteredIdContributions` empty-field filter) that the prior `apPadded` assertion already proves correct; CI will validate.

### CI

All actually-required checks are green on `d26e5a9` (50+ jobs across GitHub Actions, CircleCI, CodeQL, Codecov, signed commits, PR title convention, linked issue, License check, Frontend unit + e2e, Backend unit + integration + rules-and-taxii, pycti 3.10 / 3.11 / 3.12). `mergeable: MERGEABLE`.

One noisy duplicate workflow run remains failing: the `push`-event copy of `OpenCTI CI` (run `26330788219`) is stuck at `actions/checkout@v6` with `fatal: could not read Username for 'https://github.com'` after three rerun attempts. The semantically-equivalent `pull_request`-event copy of the same workflow (run `26330788979`) ran every job successfully on the same SHA, so this is an infrastructure flake on the duplicate run, not a real regression — `mergeable: MERGEABLE` confirms GitHub doesn't treat it as a required-check failure.

### Remaining blocker

`mergeStateStatus: BLOCKED` with `reviewDecision: REVIEW_REQUIRED`. I did not run `gh pr review --approve` because I'm the PR author and GitHub disallows self-approval — this PR still needs a maintainer approval to merge.

**Comment by SamuelHassine** [comment]
@aHenryJard All good?
